File: Bitmasking_for_Sets_states/Q3. Single Number II/Q3_SingleNumberII.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ones = 0  # stores bits seen once
        twos = 0  # stores bits seen twice
        
        # loop through each number in the list
        for num in nums:
            ones = (ones ^ num) & ~twos  # update bits that appeared once
            twos = (twos ^ num) & ~ones  # update bits that appeared twice
        
        return ones  # bits that appeared only once form the unique number

File: Bitmasking_for_Sets_states/Q3. Single Number II/Q3_SingleNumberII.java
class Solution {
    public int singleNumber(int[] nums) {
        int ones = 0;  // to store bits that have appeared once
        int twos = 0;  // to store bits that have appeared twice
        
        // Go through each number
        for (int num : nums) {
            ones = (ones ^ num) & ~twos;  // update ones: add current bit if not in twos
            twos = (twos ^ num) & ~ones;  // update twos: add current bit if not in ones
        }
        
        // The bits that appeared once represent the single number
        return ones;
    }
}
